tests/Integration/Behaviour/Features/Context/Domain/ShipmentFeatureContext.php

```diff
@@ -0,0 +1,98 @@
+ */
+
+declare(strict_types=1);
+
+namespace Tests\Integration\Behaviour\Features\Context\Domain;
+
+use Behat\Gherkin\Node\TableNode;
+use Cart;
+use PHPUnit\Framework\Assert;
+use PrestaShop\PrestaShop\Core\Domain\Shipment\Query\GetOrderShipments;
+use RuntimeException;
+use Tests\Integration\Behaviour\Features\Context\SharedStorage;
+
+class ShipmentFeatureContext extends AbstractDomainFeatureContext
+{
+    /**
+     * @Then the order :orderReference should have the following shipments:
+     *
+     * @param string $orderReference
+     * @param TableNode $table
+     *
+     * @throws RuntimeException
+     */
+    public function verifyOrderShipment(string $orderReference, TableNode $table)
+    {
+        $data = $table->getRowsHash();
+        $orderId = SharedStorage::getStorage()->get($orderReference);
+        $carrierId = SharedStorage::getStorage()->get($data['id_carrier']);
+        // get address id from cart
+        $addressId = (new Cart(SharedStorage::getStorage()->get($data['id_address'])))->id_address_delivery;
+
+        $shipments = $this->getQueryBus()->handle(
+            new GetOrderShipments($orderId)
+        );
+
+        if (count($shipments) === 0) {
+            $msg = 'Order [' . $orderId . '] has no shipments';
+            throw new RuntimeException($msg);
+        }
+
+        foreach ($shipments as $shipment) {
+            if ($shipment->getOrderId() !== $orderId) {
+                throw new RuntimeException('Shipment [' . $shipment->getId() . '] does not belong to order [' . $orderId . ']');
+            }
+
+            Assert::assertEquals($shipment->getTrackingNumber(), $data['tracking_number']);
+            Assert::assertEquals($shipment->getCarrierId(), $carrierId);
+            Assert::assertEquals($shipment->getAddressId(), $addressId);
+            Assert::assertEquals($shipment->getShippingCostTaxExcluded(), $data['shipping_cost_tax_excl']);
+            Assert::assertEquals($shipment->getShippingCostTaxIncluded(), $data['shipping_cost_tax_incl']);
+            SharedStorage::getStorage()->set($data['shipment'], $shipment);
+        }
+    }
+
+    /**
+     * @Then the shipment :shipmentReference should have the following products:
+     *
+     * @param string $shipmentReference
+     * @param TableNode $table
+     */
+    public function verifyShipmentProducts(string $shipmentReference, TableNode $table)
+    {
+        $data = $table->getColumnsHash();
+        $shipment = SharedStorage::getStorage()->get($shipmentReference);
+
+        $products = $shipment->getProducts();
+
+        Assert::assertEquals($products[0]->getQuantity(), (int) $data[0]['quantity']);
+        Assert::assertEquals($products[1]->getQuantity(), (int) $data[1]['quantity']);
+        Assert::assertEquals($products[0]->getOrderDetailId(), (int) $data[0]['order_detail']);
+        Assert::assertEquals($products[1]->getOrderDetailId(), (int) $data[1]['order_detail']);
```

Comment:
Here you should use some order detail references as well not the IDs directly the IDs are very volatile since they're based on auto-increment You cannot guaranty that they will always keep the same values 8 and 9, if another scenario (in the same feature file or from a previous one) also create some order details then the whole IDs could be offset and it would mean adapting these hard coded values all the time

That's why we store the generated IDs with string references and use those references to compare values and rely on dynamic IDs behind the scene

Reply:
I looked into the existing order scenarios quickly, but I don't see some place where actually reference the Order Detail Ids Seems like most of the steps are based on the product rather than the order detail itself So sorry you lack existing steps to handle this properly and easily

But I have another idea that should do the trick, at least for now in future PRs we may need to introduce the mapping for OrderDetails as well, you could add on the `OrderShipmentProduct` some additional data that you're probably gonna need when you will display the list anyway like:
- product name
- product ID: maybe I'm not sure about this one, it depends if some links to the product are needed in this list? if not then we shouldn't add fields in our implementation for test purpose only